### ISSUE TITLE
Component with incorrect parameter name emits no warning / error

### ISSUE THREAD

**Issue by sylvainonweb** [dotnet/aspnetcore#45667]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

If I create a component with a "Text" parameter like this  
```
<div>@Text</div>

@code {
    [Parameter]
    public string? Text { get; set; }
}
```

and put it in a page like this  
```
<MyComponent Text="a valid parameter name"></MyComponent>
<MyComponent TextWithAnInvalidParameterName="a invalid parameter name"></MyComponent>
```

If I rebuild the application, no warning / error prevents me that the application is broken.
Indeed, if I launch the application, it crashes because the component doesn't contain parameter "TextWithAnInvalidParameterName"

In my application (about 300 screens), I had a component that I refactor recently and forgot to replace in one place.  
=> same matter, application crashes.

### Expected Behavior

Visual Studio rebuild must not succeed because the application is broken

### Steps To Reproduce

Repo can be found here : https://github.com/sylvainonweb/BlazorRepo

### .NET Version

6.0.401

**Comment by javiercn** [comment]
@sylvainonweb thanks for contacting us.

There is an attribute you can use for this which is `[EditorRequired]`

**Comment by sylvainonweb** [comment]
@javiercn 
In documentation, [EditorRequired] definition is :
> Specifies that the component parameter is required to be provided by the user when authoring it in the editor

This seems to be really different of what I need. Indeed, I don't want to be forced to pass a parameter value for each of the component' parameters, I just want the IDE to prevent me if parameter doesn't exist in my component.

You say "In general, parameters are passed at runtime". 
On the contrary, I would say, "in general, parameters are known at compile time (even if parameters values are passed at runtime)". 

The fact that Blazor catch unmatched parameters is probably great for some people but is really a mess for person like me that want to trust their compiler.

=> adding an attribute on component to tell that unmatched parameters are forbidden could be a solution no ?

**Comment by sylvainonweb** [comment]
@javiercn I found an issue I created 2 years ago about the same subject (https://github.com/dotnet/razor/issues/7743) where 
@SteveSandersonMS proposed to add a strict parameters mode.

**Comment by sylvainonweb** [comment]
@javiercn IJust to be sure to understand

if I have a component like this 

```
<div>@Parameter1</div>
...
<div>@Parameter10</div>

@code {
    [Parameter, EditorRequired] public string? Parameter1 { get; set; }
    [Parameter, EditorRequired] public string? Parameter2 { get; set; }
    [Parameter, EditorRequired] public string? Parameter3 { get; set; }
    [Parameter, EditorRequired] public string? Parameter4 { get; set; }
    [Parameter, EditorRequired] public string? Parameter5 { get; set; }
    [Parameter, EditorRequired] public string? Parameter6 { get; set; }
    [Parameter, EditorRequired] public string? Parameter7 { get; set; }
    [Parameter, EditorRequired] public long? Parameter8 { get; set; }
    [Parameter, EditorRequired] public long? Parameter9 { get; set; }
    [Parameter, EditorRequired] public long? Parameter10 { get; set; }
}
```

If I had this component to a page, I wil have to write something like this

```
<MyComponent
    Parameter1=@null
    Parameter2="2"
    Parameter3="3"
    Parameter4="4"
    Parameter5="5"
    Parameter6="6"
    Parameter7="7"
    Parameter8=@null
    Parameter9=9
    Parameter10=10
></MyComponent>
```

If I want to be sure that Visual Studio can be able to detect that a parameter doesn't exist anymore in a component , I will have to :
* declare all my parameters with EditorRequired attribute
* pass parameters value to the MyComponent for all declared parameters **even if all my parameters are nullable**

=> this is the solution that you chose ?

**Comment by javiercn** [comment]
@sylvainonweb seems that we might not understood your question correctly.

`[EditorRequired]` gives a hint to the editor that the caller needs to pass in a parameter value (whatever value that is, including null).

Are you asking for the opposite? To emit a warning/error when there is a parameter passed in that does not match any parameter defined on the component?

Normally when additional parameters are passed in, they are just ignored. In the most common case where you have a typo in your component, you are going to get an error to point out the typo if you are using EditorRequired.

In general, the problem in this area is that we can't make guarantees about a parameter being required because the parameters passed to a component can be decided at runtime and the component might choose to interpret the parameters as it sees fit.

For example, I am moderately sure that [splatting](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/?view=aspnetcore-7.0#attribute-splatting-and-arbitrary-parameters) disables the editor check as it is impossible for us in that situation to know the parameter set.

I also believe that there is a runtime check that throws an exception when a parameter can't be assigned to a property, but not the other way around.

**Comment by javiercn** [comment]
@sylvainonweb fair enough.

That said, we think that most scenarios are covered by `[EditorRequired]` since the only scenario this would cover would be any additional and unknown or optional parameter.

I'll move the issue to the backlog to see if other people feel this is still a problem for them and we'll make a decision in the future.

**Comment by sylvainonweb** [comment]
@javiercn 

> That said, we think that most scenarios are covered by [EditorRequired] since the only scenario this would cover would be any additional and unknown or optional parameter.
For sure, this is the reason of the issue : not discover the error at runtime

> I'll move the issue to the backlog to see if other people feel this is still a problem for them and we'll make a decision in the future.
Event if I'm sad, I understand. What I don't really understand is how other developpers can live with this. If someone can explain me :)

Is it normal that this issue has the "Status:Resolved" tag?

**Comment by javiercn** [comment]
@sylvainonweb no, that was my mistake. I have removed it.

**Comment by begerard** [comment]
Hi @javiercn,

I just got hit in production by this, during refactoring I removed a parameter in a component and I was expecting for VS to give an error were it was used to adapt where it was used. Like it would do in Winform, WPF, Xamarin...

Unfortunately it don't, and one usage of this component go through QA undetected and crash in production. I would they say it was bound to happen :)

The Strict mode proposed by @SteveSandersonMS https://github.com/dotnet/razor/issues/7743 would have prevent that, and also speed up the refactoring in general, but the issue is closed. Is it possible reopen it, or should I make a new Feature issue from it? And more important reevaluate it as 3+ years passed since then so productivity feature may be more acceptable now!

**Comment by mohaaron** [comment]
Hi all,

I know I'm late to the conversation but razor pages get compiled into classes right? And if I was to write a class with a property name and then try to use a different property name the compile would throw a build error. This seems pretty straight forward to me and how I would expect razor component parameter names to work as well.

Just wanted to add my two cents. Thanks for listening.

**Comment by gulbanana** [comment]
I've also just experienced an error in production to this issue. A razor component library contained a parameter "IsReadonly"; it was changed to IsReadOnly; an application depending on the library updated and failed at runtime. The change was described in release notes but the application developer didn't realise they were using that parameter anywhere. 

In a C# environment where you expect breaking changes to be compiler errors, it's weird to not have this safety net - and yet, splatting is useful functionality. Perhaps we could have a way to mark components as opting in to or out of splatting?
